core/src/main/java/io/questdb/cutlass/line/tcp/LineTcpParser.java

```diff
@@ -0,0 +1,635 @@
+                        bufAt += 1;
+                        // parse quoted value
+                        if (!prepareQuotedEntity(bufAt - 1, bufHi)) {
+                            // parsing not successful
+                            if (errorCode == ErrorCode.INVALID_FIELD_VALUE_STR_UNDERFLOW) {
+                                // need more data
+                                return ParseResult.BUFFER_UNDERFLOW;
+                            }
+                            // invalid character sequence in the quoted value or EOL found
+                            return ParseResult.ERROR;
+                        }
+                        errorCode = ErrorCode.NONE;
+                        nQuoteCharacters = 0;
+                        bufAt += 1;
+                        break;
+                    } else if (isQuotedFieldValue) {
+                        return ParseResult.ERROR;
+                    }
+
+                default:
+                    appendByte = true;
+                    nextValueCanBeOpenQuote = false;
+                    break;
+            }
+
+            if (appendByte) {
+                // If there is escaped character, like \" or \\ then the escape slash has to be excluded
+                // from the result key / value.
+                // shift copy current byte back
+                if (nEscapedChars > 0) {
+                    Unsafe.getUnsafe().putByte(bufAt - nEscapedChars, b);
+                }
+                bufAt++;
+            }
+        }
+        return ParseResult.BUFFER_UNDERFLOW;
+    }
+
+    public void shl(long shl) {
+        bufAt -= shl;
+        entityLo -= shl;
+        measurementName.shl(shl);
+        charSeq.shl(shl);
+        for (int i = 0; i < nEntities; i++) {
+            entityCache.getQuick(i).shl(shl);
+        }
+    }
+
+    public ParseResult skipMeasurement(long bufHi) {
+        assert bufAt != 0 && bufHi >= bufAt;
+        while (bufAt < bufHi) {
+            byte b = Unsafe.getUnsafe().getByte(bufAt);
+            if (b == (byte) '\n' || b == (byte) '\r') {
+                return ParseResult.MEASUREMENT_COMPLETE;
+            }
+            bufAt++;
+        }
+        return ParseResult.BUFFER_UNDERFLOW;
+    }
+
+    public void startNextMeasurement() {
+        bufAt++;
+        nEscapedChars = 0;
+        isQuotedFieldValue = false;
+        entityLo = bufAt;
+        tagsComplete = false;
+        nEntities = 0;
+        currentEntity = null;
+        entityHandler = entityTableHandler;
+        timestamp = NULL_TIMESTAMP;
+        errorCode = ErrorCode.NONE;
+        nQuoteCharacters = 0;
+        scape = false;
+        nextValueCanBeOpenQuote = false;
+        hasNonAscii = false;
+    }
+
+    private boolean expectEndOfLine(byte endOfEntityByte, long bufHi) {
+        assert endOfEntityByte == '\n';
+        return true;
+    }
+
+    private boolean expectEntityName(byte endOfEntityByte, long bufHi) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 18. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 325-345 into a separate method.